Fix CDRIVER3441 - document mongoc_handshake_data_append and alias as … (#852)
mongodb/mongo-c-driver: src/libmongoc/doc/mongoc_handshake.rst

:man_page: mongoc_handshake_data_append

mongoc_handshake_data_append()

> =========================================

Reviewer: IIRC sphinx warns on mismatched underlines:

```suggestion
==============================
```

Synopsis
--------

.. code-block:: c

   bool
   mongoc_handshake_data_append (const char *driver_name,
                                 const char *driver_version,
                                 const char *platform);

Appends the given strings to the handshake data for the underlying C Driver.

Description
-----------

This function is intended for use by drivers which wrap the C Driver.
Calling this function will store the given strings as handshake data about
the system and driver by appending them to the handshake data for the
underlying C Driver. These values, along with other handshake data collected
during mongoc_init will be sent to the server as part of the initial
connection handshake in the "client" document. This function must not be

> called more than once, or after a handshake has been initiated.

Reviewer: I suggest clarifying the handshake is initiated when server monitoring begins. The [connection pooling](http://mongoc.org/libmongoc/current/connection-pooling.html) page gives additional background into the distinction of singled / pooled monitoring. Feel free to reword the suggestion.

```suggestion
called more than once, or after server monitoring begins. For a single-threaded :symbol:`mongoc_client_t`, server monitoring begins on the first operation requiring a server. For a :symbol:`mongoc_client_pool_t`, server monitoring begins on the first call to `:symbol:`mongoc_client_pool_pop`.
```

The passed in strings are copied, and don't have to remain valid after the

> call to mongoc_handshake_data_append(). The driver may store truncated

Reviewer: ```suggestion
call to :symbol:`mongoc_handshake_data_append()`. The driver may store truncated
```

To add a link.

versions of the passed in strings.

> Note:

Reviewer: Oh, good idea noting this! 

RST provides a `note` directive to indent and format. This may be applicable here.

```suggestion
.. note::
```

  This function allocates memory, and therefore caution should be used when

>   using this in conjunction with bson_mem_set_vtable. If this function is

Reviewer: I suggest adding `:symbol:` links here on the references functions listed here.

  called before bson_mem_set_vtable, then bson_mem_restore_vtable must be
  called before calling mongoc_cleanup. Failure to do so will result in
  memory being freed by the wrong allocator.

Parameters
----------

* ``driver_name``: An optional string storing the name of the wrapping driver
* ``driver_version``: An optional string storing the version of the wrapping driver.
* ``platform``: An optional string storing any information about the current platform, for example configure options or compile flags.

Returns
-------

``true`` if the given fields are set successfully. Otherwise, it returns ``false`` and logs an error.

The default handshake data the driver sends with "hello" looks something
like:

.. code-block:: c

 client: {
   driver: {
     name: "mongoc",
     version: "1.5.0"
   },
   os: {...},
   platform: "CC=gcc CFLAGS=-Wall -pedantic"
 }

If we call

>   mongoc_handshake_data_append ("phongo", "1.1.8", "CC=clang")

Reviewer: ```suggestion
  mongoc_handshake_data_append ("phongo", "1.1.8", "CC=clang / ")
```

See my comment below, but I believe the `platform` argument is prepended without additional text.

Author: The document (from the code comments) does say "something like", but it's better to be more rather than less accurate!

> and it returns true, the driver sends handshake data like:

Reviewer: This produces a warning triggering a failure in the `make-release-archive` task: https://spruce.mongodb.com/task/mongo_c_driver_releng_make_release_archive_patch_d2838038aa7d1808b872ea7bd39df33db5bba364_613294890305b906ebc5c3df_21_09_03_21_32_59/logs?execution=0

You can test the documentation build locally with:
```
$ cd /path/to/mongo-c-driver
$ cd src/libmongoc/doc
$ sphinx-build -W -b html . html
```

.. code-block::
 client: {
   driver: {
     name: "mongoc / phongo",
     version: "1.5.0 / 1.1.8"
   },
   os: {...},

>    platform: "CC=gcc CFLAGS=-Wall -pedantic / CC=clang"

Reviewer: ```suggestion
   platform: "CC=clang / CC=gcc CFLAGS=-Wall -pedantic"
```

From [this test](https://github.com/kevinAlbs/c-bootstrap/blob/90f99b57487339c3c8ac75fad259d6da1ade5e94/investigations/handshake_append.3441/main.c#L28), the current behavior prefixes and does not prepend any additional text.

:man_page: mongoc_add_driver_info

mongoc_add_driver_info()
=========================================

Synopsis
--------

.. code-block:: c

   bool

>    mongoc_add_driver_info (const char *driver_name,

Reviewer: Apologies for back-tracking. But given @alcaeus's recent feature request of CDRIVER-4142, would it be preferable to hold off on the alias, and reduce the scope of CDRIVER-3441 to documenting `mongoc_handshake_data_append`?

Author: That's a good question! I actually was wondering a bit about those semantics, myself. I'm inclined to agree with your intuition as well: the motivation for /this/ PR was to provide a name that fits in better with our extant conventions, as well as document a previously undocumented function. If we'd like the "new" name to have the "new" semantics, I think that would make sense: it doesn't break old code, and the funky name should actually partially alert callers to some funky behavior. Perhaps @alcaeus' cautionary notes should be adapted into the documentation as well. 

If you both think this is a good idea, I'm for it.

Author: Note: The changes in this PR still seem generally useful to me-- I the change would be that I remove mongo_add_driver_info() and change the references to it back to mongoc_handshake_data_append().

Author: Updated as per discussion; OR we can revert the change entirely.

Reviewer: I agree with removing `mongoc_add_driver_info()` in favor of only documenting `mongoc_handshake_data_append()`. I have updated the description of CDRIVER-3441.
